text/0000-ternary.md

```diff
@@ -0,0 +1,238 @@
+
+# Summary
+
+In this RFC we introduce both the API and implementation of the `bee-ternary`
+crate, a general-purpose ternary manipulation crate written in Rust that may be
+used to handle ternary data in the IOTA ecosystem.
+
+# Motivation
+
+Ternary is a fundamental part of the IOTA ecosystem and is used throughout, even
+for fundamental features like wallet addresses and transaction identification.
+
+Manipulating ternary in an ergonomic manner on top of existing binary-driven
+code can require complexity and, up until now, comes either with a lot of
+boilerplate code, implementations that are difficult to verify as correct, or a
+lack of features.
+
+`bee-ternary` seeks to become the canonical ternary crate in the Rust IOTA
+ecosystem by providing an API that is efficient, ergonomic, and featureful all
+at once.
+
+We hope that `bee-ternary` will allow IOTA developers to more easily write code
+that works with ternary and allow the simplification of existing codebases
+through use of the API.
+
+Broadly, there are 3 main benefits things that `bee-ternary` aims to provide
+over use-specific implementations of ternary manipulation code.
+
+- **Ergonomics**: The API should be trivial to use correctly and should
+naturally guide developers towards correct and efficient solutions.
+
+- **Features**: The API should provide a powerful set of core features that can
+be used together to cover most developer requirements. Examples of such features
+include multiple encoding schemes, binary/ternary conversion, (de)serialization,
+and tryte string de/encoding.
+
+- **Performance**: The API should allow 'acceptably' efficient manipulation of
+ternary data. This is difficult to ruggedly define, and implementation details
+may change later, but broadly we intend to introduce an API that doesn't
+inherently inhibit high performance through poor design choices.
+
+# Detailed design
+
+`bee-ternary` is designed to be extensible and is built on top of a handful of
+powerful abstraction types. Below we list the features of the API and summarise
+a few of the core API objects.
+
+## Features
+
+- Efficient manipulation of ternary buffers (trits and trytes)
+- Multiple encoding schemes
+- Extensible design that allows it to sit on top of existing data structures,
+avoiding unnecessary allocation and copying
+- A plethora of utility functions to allow for easy manipulation of ternary data
+- Zero-cost conversion between trit and tryte formats
+
+## Key Types & Traits
+
+### `Trits`
+
+```rust
+struct Trits<T: RawEncoding> { ... }
+```
+
+- Analogous to `str` or `[T]`
+- Unsized type, represents a buffer of trits of a specified encoding
+- Most commonly used from behind a reference (as with `&str` and `&[T]`),
+representing a 'trit slice'
+- Can be created from a variety of types such as`TritBuf` and `[i8]` with
+minimal overhead
+- Sub-slices can be created from trit slices at a per-trit level
+
+### `TritBuf`
+
+```rust
+struct TritBuf<T: RawEncodingBuf> { ... }
```

Comment:
There's a lot of documentation that would come with that, and we'd need to provide many impl blocks (since methods differ for different type parameters).

Also, the internal field(s) are an implementation detail and should not be exposed to the API user.

Reply:
hmmm.... This RFC is supposed to guide the implementer (which in general is a different person than the RFC author). It's not written for the API user. I think there is no need to provide `impl` blocks, but shouldn't the implementer know, how this type is supposed to realized internally?

Reply:
Those field definitions might well change later on. Besides, the fields don't actually relate to the implementation all that much since they rely on unsized type tricks to work. I really don't think it would be useful to do more than pointing to the actual implementation given the complexities associated with it.

Reply:
Alright, that works for me. 

For the record: I wouldn't be too much concerned with your first statement, that field definitions might change. That's true, but I think RFC are not supposed to be necessarily set in stone, and can be updated later on. If the changes are too significant, then an additional improvement or even replacement RFC should be the way to go. :smile:

Reply:
Again,,,I think my concerns were addressed. Thanks.